sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c

```diff
@@ -0,0 +1,304 @@
+#define ID_SCOPE_ENV "AZ_IOT_ID_SCOPE"
+
+// Device information
+#define REGISTRATION_ID_ENV "AZ_IOT_REGISTRATION_ID"
+// AZ_IOT_DEVICE_X509_CERT_PEM_FILE is the path to a PEM file containing the device certificate and
+// key as well as any intermediate certificates chaining to an uploaded group certificate.
+
+#define DEVICE_X509_CERT_PEM_FILE "AZ_IOT_DEVICE_X509_CERT_PEM_FILE"
+
+static char global_provisioning_endpoint[256] = { 0 };
+static char id_scope[16] = { 0 };
+static char registration_id[256] = { 0 };
+static char x509_cert_pem_file[256] = { 0 };
+
+static az_iot_provisioning_client provisioning_client;
+static MQTTClient mqtt_client;
+
+static az_result read_configuration_entry(
+    const char* name,
+    const char* env_name,
+    char* default_value,
+    bool hide_value,
+    az_span buffer,
+    az_span* out_value)
+{
+  printf("%s = ", name);
+  char* env = getenv(env_name);
+
+  if (env != NULL)
+  {
+    printf("%s\n", hide_value ? "***" : env);
+    *out_value = az_span_init(az_span_ptr(buffer), 0, az_span_capacity(buffer));
+    AZ_RETURN_IF_FAILED(az_span_copy(*out_value, az_span_from_str(env), out_value));
+  }
+  else if (default_value != NULL)
+  {
+    printf("%s\n", default_value);
+    AZ_RETURN_IF_FAILED(az_span_copy(*out_value, az_span_from_str(default_value), out_value));
+  }
+  else
+  {
+    printf("(missing) Please set the %s environment variable.\n", env_name);
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+static az_result read_configuration()
+{
+  az_span endpoint_span = AZ_SPAN_LITERAL_FROM_BUFFER(global_provisioning_endpoint);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "Global Device Endpoint",
+      GLOBAL_PROVISIONING_ENDPOINT_ENV,
+      GLOBAL_PROVISIONING_ENDPOINT_DEFAULT,
+      false,
+      endpoint_span,
+      &endpoint_span));
+
+  az_span id_scope_span = AZ_SPAN_LITERAL_FROM_BUFFER(id_scope);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "ID_Scope", ID_SCOPE_ENV, NULL, false, id_scope_span, &id_scope_span));
+
+  az_span registration_id_span = AZ_SPAN_LITERAL_FROM_BUFFER(registration_id);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "Registration ID",
+      REGISTRATION_ID_ENV,
+      NULL,
+      false,
+      registration_id_span,
+      &registration_id_span));
+
+  az_span temp = AZ_SPAN_LITERAL_FROM_BUFFER(x509_cert_pem_file);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      "X509 Certificate PEM Store File", DEVICE_X509_CERT_PEM_FILE, NULL, false, temp, &temp));
+
+  AZ_RETURN_IF_FAILED(az_iot_provisioning_client_init(
+      &provisioning_client, endpoint_span, id_scope_span, registration_id_span, NULL));
+
+  return AZ_OK;
+}
+
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  UNUSED(context);
+  UNUSED(topicLen);
+
+  int i;
+  char* payloadptr;
+
+  printf("Message arrived\n");
+  printf("     topic: %s\n", topicName);
```

Comment:
Can we print this topic with length? Reason of bubbling length is that people think it like buffer.


printf("     topic: %.*s\n", topicLen, topicName);

Reply:
This will be removed: no point in printing out every message during provisioning. (I left this in by mistake.)
Paho has good logging that should display the info you need. If that's not true, I'll leave it in and add len.

Reply:
You could have used %.*s to bound the printf 

---
In reply to: [406367096](https://github.com/Azure/azure-sdk-for-c/pull/547#discussion_r406367096) [](ancestors = 406367096)